tests/test_reply_function_sync.js

```diff
@@ -48,152 +33,199 @@ test('get with reply callback returning object', async t => {
-  scope.done()
-})
-
-test('reply headers work with function', async t => {
-  const scope = nock('http://example.com')
-    .get('/')
-    .reply(200, () => 'ABC', { 'X-My-Headers': 'My custom header value' })
+    .reply(200, () => 123)
 
-  const { headers } = await got('http://example.com/')
-
-  t.equivalent(headers, { 'x-my-headers': 'My custom header value' })
+  const { body } = await got('http://example.test')
+  t.equal(body, '123')
   scope.done()
 })
 
-// Skipped because https://github.com/nock/nock/issues/1222
+// The observed behavior is that this returns a 123 status code.
+//
+// The expected behavior is that this should either throw an error or reply
+// with 200 and the JSON-stringified '[123]'.
 test(
-  'get with reply callback returning default statusCode without body',
+  'reply with status code and function returning array',
   { skip: true },
-  t => {
-    nock('http://replyheaderland')
+  async t => {
+    const scope = nock('http://example.test')
       .get('/')
-      .reply((uri, requestBody) => [401])
-
-    http.get(
-      {
-        host: 'replyheaderland',
-        path: '/',
-        port: 80,
-      },
-      res => {
-        res.setEncoding('utf8')
-        t.equal(res.statusCode, 200)
-        res.on('data', data => {
-          t.equal(data, '[401]')
-          res.once('end', t.end.bind(t))
-        })
-      }
-    )
+      .reply(200, () => [123])
+
+    const { body } = await got('http://example.test')
+    t.equal(body, '[123]')
+    scope.done()
   }
 )
 
-test('get with reply callback returning callback without headers', async t => {
-  const scope = nock('http://example.com')
-    .get('/')
-    .reply(() => [401, 'This is a body'])
+test('reply function with string body using POST', async t => {
+  const exampleRequestBody = 'key=val'
+  const exampleResponseBody = 'foo'
 
-  await assertRejects(got('http://example.com/'), err => {
-    t.equal(err.statusCode, 401)
-    t.equal(err.body, 'This is a body')
-    return true
-  })
+  const scope = nock('http://example.test')
+    .post('/endpoint', exampleRequestBody)
+    .reply(404, () => exampleResponseBody)
+
+  await assertRejects(
+    got.post('http://example.test/endpoint', {
+      body: exampleRequestBody,
+    }),
+    ({ statusCode, body }) => {
+      t.equal(statusCode, 404)
+      t.equal(body, exampleResponseBody)
+      return true
+    }
+  )
   scope.done()
 })
 
-test('post with reply callback, uri, and request body', async t => {
-  const input = 'key=val'
+test('reply function receives the request URL and body', async t => {
+  t.plan(3)
 
-  const scope = nock('http://example.com')
-    .post('/echo', input)
-    .reply(200, (uri, body) => ['OK', uri, body].join(' '))
+  const exampleRequestBody = 'key=val'
+
+  const scope = nock('http://example.test')
+    .post('/endpoint', exampleRequestBody)
+    .reply(404, (uri, requestBody) => {
+      t.equal(uri, '/endpoint')
+      t.equal(requestBody, exampleRequestBody)
```

Comment:
Is there anything that asserts these two lines get executed?

Reply:
Fortunately, yes: the `t.plan(3)` means exactly three tap asserts have to be called for the test to pass.